### DISCUSSION TITLE
Please clarify how throwing inside `subscribe()` works

### DISCUSSION THREAD

**Discussion by pastelmind** [pmndrs/valtio#842] [General]
Currently, `subscribe()` *consumes* errors thrown inside the callback:

```js
// There is no way to catch this error, other than listening to the global
// 'unhandledrejection' event
subscribe(store, () => { throw new Error() })

// This will not throw
store.value = 1
```

This is because `subscribe()` does not handle rejections here:

https://github.com/pmndrs/valtio/blob/26caa885050b5c9ac8c80c90e216599936e925f8/src/vanilla.ts#L378-L383

If a synchronously subscribed callback throws, it *does* cause the offending mutation to throw. But it will also cause all subscribers registered after it to be skipped:

```js
subscribe(store, () => { throw new Error() }, true);
subscribe(store, () => console.log('I am skipped'), true);
subscribe(store, () => console.log('So am I'));

assert.throws(() => {
  // This will throw
  store.value = 1
})
```

This is because all subscribers are invoked synchronously here:

https://github.com/pmndrs/valtio/blob/26caa885050b5c9ac8c80c90e216599936e925f8/src/vanilla.ts#L176

I just want to know if this is intended. I don't care if it is fixed or kept this way. But it would be beneficial if this behavior is documented somewhere.

**Comment by dai-shi** [comment]
In the valtio mental model, you can't stop the mutation. subscribe is called to notify that the mutation happened after the mutation.

I think we should say, throwing in the subscription callbacks leads to an unexpected behavior.
